File: tests/fixtures/sample.go
// SPDX-FileCopyrightText: 2025 Knitli Inc.
// SPDX-License-Identifier: MIT OR Apache-2.0

// Package fixtures provides sample Go code for testing chunker functionality.
package fixtures

import (
    "fmt"
    "sync"
    "time"
)

// Processor defines the interface for data processing operations.
type Processor interface {
    Process(item DataItem) (DataItem, error)
    ProcessBatch(items []DataItem) ([]DataItem, error)
}

// DataItem represents a single data element with metadata.
type DataItem struct {
    ID        string
    Value     int
    Timestamp time.Time
    Metadata  map[string]string
}

// Cache provides thread-safe caching with TTL support.
type Cache struct {
    mu      sync.RWMutex
    storage map[string]cacheEntry
    ttl     time.Duration
}

type cacheEntry struct {
    value     interface{}
    expiresAt time.Time
}

// NewCache creates a new cache with specified TTL.
func NewCache(ttl time.Duration) *Cache {
    return &Cache{
        storage: make(map[string]cacheEntry),
        ttl:     ttl,
    }
}

// Get retrieves a value from cache if not expired.
func (c *Cache) Get(key string) (interface{}, bool) {
    c.mu.RLock()
    defer c.mu.RUnlock()

    entry, exists := c.storage[key]
    if !exists {
        return nil, false
    }

    if time.Now().After(entry.expiresAt) {
        return nil, false
    }

    return entry.value, true
}

// Set stores a value in cache with TTL.
func (c *Cache) Set(key string, value interface{}) {
    c.mu.Lock()
    defer c.mu.Unlock()

    c.storage[key] = cacheEntry{
        value:     value,
        expiresAt: time.Now().Add(c.ttl),
    }
}

// DefaultProcessor implements the Processor interface.
type DefaultProcessor struct {
    cache      *Cache
    multiplier int
}

// NewDefaultProcessor creates a processor with caching.
func NewDefaultProcessor(multiplier int) *DefaultProcessor {
    return &DefaultProcessor{
        cache:      NewCache(5 * time.Minute),
        multiplier: multiplier,
    }
}

// Process transforms a single data item.
func (p *DefaultProcessor) Process(item DataItem) (DataItem, error) {
    if item.ID == "" {
        return DataItem{}, fmt.Errorf("missing item ID")
    }

    item.Value *= p.multiplier
    item.Timestamp = time.Now()

    return item, nil
}

// ProcessBatch processes multiple items concurrently.
func (p *DefaultProcessor) ProcessBatch(items []DataItem) ([]DataItem, error) {
    results := make([]DataItem, len(items))
    var wg sync.WaitGroup

    for i, item := range items {
        wg.Add(1)
        go func(idx int, it DataItem) {
            defer wg.Done()
            result, _ := p.Process(it)
            results[idx] = result
        }(i, item)
    }

    wg.Wait()
    return results, nil
}
─────┴──────────────────────────────────────────────────────────────────────────

File: tests/fixtures/sample.js
// SPDX-FileCopyrightText: 2025 Knitli Inc.
// SPDX-License-Identifier: MIT OR Apache-2.0

/**
 * Sample JavaScript module for testing chunker functionality.
 * Contains nested functions, closures, and class definitions.
 */

class TaskQueue {
  constructor(maxConcurrency = 5) {
    this.maxConcurrency = maxConcurrency;
    this.queue = [];
    this.running = 0;
  }

  async enqueue(task) {
    return new Promise((resolve, reject) => {
      this.queue.push({ task, resolve, reject });
      this.processNext();
    });
  }

  async processNext() {
    if (this.running >= this.maxConcurrency || this.queue.length === 0) {
      return;
    }

    this.running++;
    const { task, resolve, reject } = this.queue.shift();

    try {
      const result = await task();
      resolve(result);
    } catch (error) {
      reject(error);
    } finally {
      this.running--;
      this.processNext();
    }
  }
}

function createDataProcessor(config) {
  const cache = new Map();
  const stats = { hits: 0, misses: 0 };

  function processItem(item) {
    const cached = cache.get(item.id);
    if (cached) {
      stats.hits++;
      return cached;
    }

    stats.misses++;
    const result = {
      id: item.id,
      value: transformValue(item.value),
      timestamp: Date.now()
    };

    cache.set(item.id, result);
    return result;
  }

  function transformValue(value) {
    return config.multiplier ? value * config.multiplier : value;
  }

  function getStats() {
    return { ...stats };
  }

  return {
    process: processItem,
    stats: getStats,
    cache
  };
}

const withRetry = (fn, maxAttempts = 3) => {
  return async (...args) => {
    let lastError;
    
    for (let attempt = 1; attempt <= maxAttempts; attempt++) {
      try {
        return await fn(...args);
      } catch (error) {
        lastError = error;
        if (attempt < maxAttempts) {
          await new Promise(resolve => setTimeout(resolve, 1000 * attempt));
        }
      }
    }
    
    throw lastError;
  };
};

module.exports = { TaskQueue, createDataProcessor, withRetry };
─────┴──────────────────────────────────────────────────────────────────────────
